File: ios/RCTPSPDFKit/PDFMemoryManager.m
#import "PDFMemoryManager.h"

@implementation PDFMemoryManager

static NSMutableSet *activeDocumentPaths;
static NSMutableDictionary *lastUsedTimes;

+ (void)initialize {
    if (self == [PDFMemoryManager class]) {
        activeDocumentPaths = [NSMutableSet new];
        lastUsedTimes = [NSMutableDictionary new];
    }
}

+ (void)registerDocument:(NSString *)path {
    dispatch_async(dispatch_get_main_queue(), ^{
        [activeDocumentPaths addObject:path];
        [lastUsedTimes setObject:[NSDate date] forKey:path];
        
        if (activeDocumentPaths.count > 1) {
            [self cleanupOldestDocument];
        }
    });
}

+ (void)cleanupDocument:(NSString *)path {
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([activeDocumentPaths containsObject:path]) {
            [activeDocumentPaths removeObject:path];
            
            for (PSPDFDocument *document in PSPDFKit.SDK.shared.documentRegistry) {
                PSPDFDocumentProvider *provider = document.documentProviders.firstObject;
                if ([provider.fileURL.path isEqualToString:path]) {
                    [document cancelFindString];
                    [document clearCache];
                    [PSPDFKit.SDK.shared.cache invalidateImagesFromDocument:document];
                }
            }
        }
    });
}

+ (void)cleanupOldestDocument {
    if (activeDocumentPaths.count == 0) return;
    
    NSString *oldestPath = nil;
    NSDate *oldestDate = [NSDate date];
    
    for (NSString *path in activeDocumentPaths) {
        NSDate *date = lastUsedTimes[path];
        if ([date compare:oldestDate] == NSOrderedAscending) {
            oldestDate = date;
            oldestPath = path;
        }
    }
    
    if (oldestPath) {
        [self cleanupDocument:oldestPath];
    }
}

+ (void)forceGlobalCleanup {
    dispatch_async(dispatch_get_main_queue(), ^{
        [PSPDFKit.SDK.shared.cache clear];
        [PSPDFKit.SDK.shared.renderManager purgeCache];
        
        NSSet *paths = [activeDocumentPaths copy];
        for (NSString *path in paths) {
            [self cleanupDocument:path];
        }
        [activeDocumentPaths removeAllObjects];
    });
}

@end 

File: ios/RCTPSPDFKit/PDFMemoryManager.swift
import Foundation
import PSPDFKit

@objc public class PDFMemoryManager: NSObject {
    
    private static var activeDocumentPaths = NSMutableSet()
    private static var lastUsedTimes = NSMutableDictionary()
    
    @objc public static func registerDocument(path: String) {
        DispatchQueue.main.async {
            activeDocumentPaths.add(path)
            lastUsedTimes.setObject(Date(), forKey: path as NSString)
            
            // If too many documents are open, force close oldest
            if activeDocumentPaths.count > 1 {
                cleanupOldestDocument()
            }
        }
    }
    
    @objc public static func cleanupDocument(path: String) {
        DispatchQueue.main.async {
            if activeDocumentPaths.contains(path) {
                activeDocumentPaths.remove(path)
                PSPDFKit.SDK.shared.documentRegistry.forEach({ document in
                    if document.documentProviders.first?.fileURL.path == path {
                        // Force document closure
                        document.cancelFindString()
                        document.clearCache()
                        PSPDFKit.SDK.shared.cache.invalidateImages(from: document)
                    }
                })
                // We'll rely on the RCTPSPDFKitManager methods to handle this
            }
        }
    }
    
    @objc public static func cleanupOldestDocument() {
        guard activeDocumentPaths.count > 0 else { return }
        
        var oldestPath: String?
        var oldestDate = Date()
        
        // Find oldest document
        for path in activeDocumentPaths {
            guard let path = path as? String,
                  let date = lastUsedTimes.object(forKey: path) as? Date else { continue }
            
            if date.compare(oldestDate) == .orderedAscending {
                oldestDate = date
                oldestPath = path
            }
        }
        
        if let path = oldestPath {
            cleanupDocument(path: path)
        }
    }
    
    @objc public static func forceGlobalCleanup() {
        DispatchQueue.main.async {
            // Clear SDK caches
            PSPDFKit.SDK.shared.cache.clear()
            PSPDFKit.SDK.shared.renderManager.purgeCache()
            PSPDFKit.SDK.shared.backgroundColor = .white
            
            // Force GC
            autoreleasepool {
                for path in activeDocumentPaths.copy() as! Set<String> {
                    cleanupDocument(path: path)
                }
                activeDocumentPaths.removeAllObjects()
            }
            
            // We'll rely on the RCTPSPDFKitManager methods to handle this
        }
    }
} 